### DISCUSSION TITLE
Build for linux-mips32 with CFLAGS="-mabi=32 -mips2 -mmxu3 -mfp64" does not work?

### DISCUSSION THREAD

**Discussion by jinxiangwzh** [openssl/openssl#21432] [Q&A] [answered]
**Openssl version**： 3.1.1
**Configure**
`./Configure linux-mips32 --cross-compile-prefix=mips-linux-gnu- --prefix=/home/jamo/workspace/protocol_example/lib CFLAGS="-mabi=32 -mips2 -mmxu3 -mfp64"`

output:
` ./Configure linux-mips32 --cross-compile-prefix=mips-linux-gnu- --prefix=/home/jamo/workspace/protocol_example/lib CFLAGS="-mabi=32 -mips2 -mmxu3 -mfp64"
Configuring OpenSSL version 3.1.1 for target linux-mips32
Using os-specific seed configuration
Created configdata.pm
Running configdata.pm
Created Makefile.in
Created Makefile

**********************************************************************
***                                                                ***
***   OpenSSL has been successfully configured                     ***
***                                                                ***
***   If you encounter a problem while building, please open an    ***
***   issue on GitHub <https://github.com/openssl/openssl/issues>  ***
***   and include the output from the following command:           ***
***                                                                ***
***       perl configdata.pm --dump                                ***
***                                                                ***
***   (If you are new to OpenSSL, you might want to consult the    ***
***   'Troubleshooting' section in the INSTALL.md file first)      ***
***                                                                ***
**********************************************************************`
**build error:**
`make -j8
/usr/bin/perl "-I." -Mconfigdata "util/dofile.pl" "-oMakefile" include/crypto/bn_conf.h.in > include/crypto/bn_conf.h
/usr/bin/perl "-I." -Mconfigdata "util/dofile.pl" "-oMakefile" include/crypto/dso_conf.h.in > include/crypto/dso_conf.h
/usr/bin/perl "-I." -Mconfigdata "util/dofile.pl" "-oMakefile" include/openssl/asn1.h.in > include/openssl/asn1.h
/usr/bin/perl "-I." -Mconfigdata "util/dofile.pl" "-oMakefile" include/openssl/asn1t.h.in > include/openssl/asn1t.h
/usr/bin/perl "-I." -Mconfigdata "util/dofile.pl" "-oMakefile" include/openssl/bio.h.in > include/openssl/bio.h
/usr/bin/perl "-I." -Mconfigdata "util/dofile.pl" "-oMakefile" include/openssl/cmp.h.in > include/openssl/cmp.h
/usr/bin/perl "-I." -Mconfigdata "util/dofile.pl" "-oMakefile" include/openssl/cms.h.in > include/openssl/cms.h
/usr/bin/perl "-I." -Mconfigdata "util/dofile.pl" "-oMakefile" include/openssl/conf.h.in > include/openssl/conf.h
/usr/bin/perl "-I." -Mconfigdata "util/dofile.pl" "-oMakefile" include/openssl/crmf.h.in > include/openssl/crmf.h
/usr/bin/perl "-I." -Mconfigdata "util/dofile.pl" "-oMakefile" include/openssl/crypto.h.in > include/openssl/crypto.h
/usr/bin/perl "-I." -Mconfigdata "util/dofile.pl" "-oMakefile" include/openssl/ct.h.in > include/openssl/ct.h
/usr/bin/perl "-I." -Mconfigdata "util/dofile.pl" "-oMakefile" include/openssl/err.h.in > include/openssl/err.h
/usr/bin/perl "-I." -Mconfigdata "util/dofile.pl" "-oMakefile" include/openssl/ess.h.in > include/openssl/ess.h
/usr/bin/perl "-I." -Mconfigdata "util/dofile.pl" "-oMakefile" include/openssl/fipskey.h.in > include/openssl/fipskey.h
/usr/bin/perl "-I." -Mconfigdata "util/dofile.pl" "-oMakefile" include/openssl/lhash.h.in > include/openssl/lhash.h
/usr/bin/perl "-I." -Mconfigdata "util/dofile.pl" "-oMakefile" include/openssl/ocsp.h.in > include/openssl/ocsp.h
/usr/bin/perl "-I." -Mconfigdata "util/dofile.pl" "-oMakefile" include/openssl/opensslv.h.in > include/openssl/opensslv.h
/usr/bin/perl "-I." -Mconfigdata "util/dofile.pl" "-oMakefile" include/openssl/pkcs12.h.in > include/openssl/pkcs12.h
/usr/bin/perl "-I." -Mconfigdata "util/dofile.pl" "-oMakefile" include/openssl/pkcs7.h.in > include/openssl/pkcs7.h
/usr/bin/perl "-I." -Mconfigdata "util/dofile.pl" "-oMakefile" include/openssl/safestack.h.in > include/openssl/safestack.h
/usr/bin/perl "-I." -Mconfigdata "util/dofile.pl" "-oMakefile" include/openssl/srp.h.in > include/openssl/srp.h
/usr/bin/perl "-I." -Mconfigdata "util/dofile.pl" "-oMakefile" include/openssl/ssl.h.in > include/openssl/ssl.h
/usr/bin/perl "-I." -Mconfigdata "util/dofile.pl" "-oMakefile" include/openssl/ui.h.in > include/openssl/ui.h
/usr/bin/perl "-I." -Mconfigdata "util/dofile.pl" "-oMakefile" include/openssl/x509.h.in > include/openssl/x509.h
/usr/bin/perl "-I." -Mconfigdata "util/dofile.pl" "-oMakefile" include/openssl/x509_vfy.h.in > include/openssl/x509_vfy.h
/usr/bin/perl "-I." -Mconfigdata "util/dofile.pl" "-oMakefile" include/openssl/x509v3.h.in > include/openssl/x509v3.h
/usr/bin/perl "-I." -Mconfigdata "util/dofile.pl" "-oMakefile" test/provider_internal_test.cnf.in > test/provider_internal_test.cnf
make depend && make _build_sw
make[1]: Entering directory '/home/jamo/workspace/protocol_example/thirdparty/openssl-3.1.1'
make[1]: Leaving directory '/home/jamo/workspace/protocol_example/thirdparty/openssl-3.1.1'
make[1]: Entering directory '/home/jamo/workspace/protocol_example/thirdparty/openssl-3.1.1'
mips-linux-gnu-gcc  -I. -Iinclude -Iapps/include  -fPIC -pthread -mabi=32 -Wa,--noexecstack -mabi=32 -mips2 -mmxu3 -mfp64 -DOPENSSL_USE_NODELETE -DOPENSSL_PIC -DOPENSSLDIR="\"/home/jamo/workspace/protocol_example/lib/ssl\"" -DENGINESDIR="\"/home/jamo/workspace/protocol_example/lib/lib/engines-3\"" -DMODULESDIR="\"/home/jamo/workspace/protocol_example/lib/lib/ossl-modules\"" -DOPENSSL_BUILDING_OPENSSL -DNDEBUG  -MMD -MF apps/lib/libapps-lib-app_libctx.d.tmp -MT apps/lib/libapps-lib-app_libctx.o -c -o apps/lib/libapps-lib-app_libctx.o apps/lib/app_libctx.c
mips-linux-gnu-gcc  -I. -Iinclude -Iapps/include  -fPIC -pthread -mabi=32 -Wa,--noexecstack -mabi=32 -mips2 -mmxu3 -mfp64 -DOPENSSL_USE_NODELETE -DOPENSSL_PIC -DOPENSSLDIR="\"/home/jamo/workspace/protocol_example/lib/ssl\"" -DENGINESDIR="\"/home/jamo/workspace/protocol_example/lib/lib/engines-3\"" -DMODULESDIR="\"/home/jamo/workspace/protocol_example/lib/lib/ossl-modules\"" -DOPENSSL_BUILDING_OPENSSL -DNDEBUG  -MMD -MF apps/lib/libapps-lib-app_params.d.tmp -MT apps/lib/libapps-lib-app_params.o -c -o apps/lib/libapps-lib-app_params.o apps/lib/app_params.c
mips-linux-gnu-gcc  -I. -Iinclude -Iapps/include  -fPIC -pthread -mabi=32 -Wa,--noexecstack -mabi=32 -mips2 -mmxu3 -mfp64 -DOPENSSL_USE_NODELETE -DOPENSSL_PIC -DOPENSSLDIR="\"/home/jamo/workspace/protocol_example/lib/ssl\"" -DENGINESDIR="\"/home/jamo/workspace/protocol_example/lib/lib/engines-3\"" -DMODULESDIR="\"/home/jamo/workspace/protocol_example/lib/lib/ossl-modules\"" -DOPENSSL_BUILDING_OPENSSL -DNDEBUG  -MMD -MF apps/lib/libapps-lib-app_provider.d.tmp -MT apps/lib/libapps-lib-app_provider.o -c -o apps/lib/libapps-lib-app_provider.o apps/lib/app_provider.c
mips-linux-gnu-gcc  -I. -Iinclude -Iapps/include  -fPIC -pthread -mabi=32 -Wa,--noexecstack -mabi=32 -mips2 -mmxu3 -mfp64 -DOPENSSL_USE_NODELETE -DOPENSSL_PIC -DOPENSSLDIR="\"/home/jamo/workspace/protocol_example/lib/ssl\"" -DENGINESDIR="\"/home/jamo/workspace/protocol_example/lib/lib/engines-3\"" -DMODULESDIR="\"/home/jamo/workspace/protocol_example/lib/lib/ossl-modules\"" -DOPENSSL_BUILDING_OPENSSL -DNDEBUG  -MMD -MF apps/lib/libapps-lib-app_rand.d.tmp -MT apps/lib/libapps-lib-app_rand.o -c -o apps/lib/libapps-lib-app_rand.o apps/lib/app_rand.c
mips-linux-gnu-gcc  -I. -Iinclude -Iapps/include  -fPIC -pthread -mabi=32 -Wa,--noexecstack -mabi=32 -mips2 -mmxu3 -mfp64 -DOPENSSL_USE_NODELETE -DOPENSSL_PIC -DOPENSSLDIR="\"/home/jamo/workspace/protocol_example/lib/ssl\"" -DENGINESDIR="\"/home/jamo/workspace/protocol_example/lib/lib/engines-3\"" -DMODULESDIR="\"/home/jamo/workspace/protocol_example/lib/lib/ossl-modules\"" -DOPENSSL_BUILDING_OPENSSL -DNDEBUG  -MMD -MF apps/lib/libapps-lib-app_x509.d.tmp -MT apps/lib/libapps-lib-app_x509.o -c -o apps/lib/libapps-lib-app_x509.o apps/lib/app_x509.c
mips-linux-gnu-gcc  -I. -Iinclude -Iapps/include  -fPIC -pthread -mabi=32 -Wa,--noexecstack -mabi=32 -mips2 -mmxu3 -mfp64 -DOPENSSL_USE_NODELETE -DOPENSSL_PIC -DOPENSSLDIR="\"/home/jamo/workspace/protocol_example/lib/ssl\"" -DENGINESDIR="\"/home/jamo/workspace/protocol_example/lib/lib/engines-3\"" -DMODULESDIR="\"/home/jamo/workspace/protocol_example/lib/lib/ossl-modules\"" -DOPENSSL_BUILDING_OPENSSL -DNDEBUG  -MMD -MF apps/lib/libapps-lib-apps.d.tmp -MT apps/lib/libapps-lib-apps.o -c -o apps/lib/libapps-lib-apps.o apps/lib/apps.c
mips-linux-gnu-gcc  -I. -Iinclude -Iapps/include  -fPIC -pthread -mabi=32 -Wa,--noexecstack -mabi=32 -mips2 -mmxu3 -mfp64 -DOPENSSL_USE_NODELETE -DOPENSSL_PIC -DOPENSSLDIR="\"/home/jamo/workspace/protocol_example/lib/ssl\"" -DENGINESDIR="\"/home/jamo/workspace/protocol_example/lib/lib/engines-3\"" -DMODULESDIR="\"/home/jamo/workspace/protocol_example/lib/lib/ossl-modules\"" -DOPENSSL_BUILDING_OPENSSL -DNDEBUG  -MMD -MF apps/lib/libapps-lib-apps_ui.d.tmp -MT apps/lib/libapps-lib-apps_ui.o -c -o apps/lib/libapps-lib-apps_ui.o apps/lib/apps_ui.c
mips-linux-gnu-gcc  -I. -Iinclude -Iapps/include  -fPIC -pthread -mabi=32 -Wa,--noexecstack -mabi=32 -mips2 -mmxu3 -mfp64 -DOPENSSL_USE_NODELETE -DOPENSSL_PIC -DOPENSSLDIR="\"/home/jamo/workspace/protocol_example/lib/ssl\"" -DENGINESDIR="\"/home/jamo/workspace/protocol_example/lib/lib/engines-3\"" -DMODULESDIR="\"/home/jamo/workspace/protocol_example/lib/lib/ossl-modules\"" -DOPENSSL_BUILDING_OPENSSL -DNDEBUG  -MMD -MF apps/lib/libapps-lib-columns.d.tmp -MT apps/lib/libapps-lib-columns.o -c -o apps/lib/libapps-lib-columns.o apps/lib/columns.c
cc1: error: '-mgp32' and '-mfp64' can only be combined if the target supports the mfhc1 and mthc1 instructions
cc1: error: '-mgp32' and '-mfp64' can only be combined if the target supports the mfhc1 and mthc1 instructions
cc1: error: '-mgp32' and '-mfp64' can only be combined if the target supports the mfhc1 and mthc1 instructions
cc1: error: '-mgp32' and '-mfp64' can only be combined if the target supports the mfhc1 and mthc1 instructions
cc1: error: '-mgp32' and '-mfp64' can only be combined if the target supports the mfhc1 and mthc1 instructions
cc1: error: '-mgp32' and '-mfp64' can only be combined if the target supports the mfhc1 and mthc1 instructions
cc1: error: '-mgp32' and '-mfp64' can only be combined if the target supports the mfhc1 and mthc1 instructions
cc1: warning: the 'mips2' architecture does not support '-mnan=2008'
cc1: warning: the 'mips2' architecture does not support '-mnan=2008'
cc1: warning: the 'mips2' architecture does not support '-mnan=2008'
cc1: warning: the 'mips2' architecture does not support '-mnan=2008'
cc1: warning: the 'mips2' architecture does not support '-mnan=2008'
cc1: error: '-mgp32' and '-mfp64' can only be combined if the target supports the mfhc1 and mthc1 instructions
cc1: warning: the 'mips2' architecture does not support '-mnan=2008'
cc1: warning: the 'mips2' architecture does not support '-mnan=2008'
cc1: warning: the 'mips2' architecture does not support '-mnan=2008'
make[1]: *** [Makefile:4317: apps/lib/libapps-lib-app_libctx.o] Error 1
make[1]: *** Waiting for unfinished jobs....
make[1]: *** [Makefile:4325: apps/lib/libapps-lib-app_params.o] Error 1
make[1]: *** [Makefile:4333: apps/lib/libapps-lib-app_provider.o] Error 1
make[1]: *** [Makefile:4341: apps/lib/libapps-lib-app_rand.o] Error 1
make[1]: *** [Makefile:4349: apps/lib/libapps-lib-app_x509.o] Error 1
make[1]: *** [Makefile:4357: apps/lib/libapps-lib-apps.o] Error 1
make[1]: *** [Makefile:4365: apps/lib/libapps-lib-apps_ui.o] Error 1
make[1]: *** [Makefile:4373: apps/lib/libapps-lib-columns.o] Error 1
make[1]: Leaving directory '/home/jamo/workspace/protocol_example/thirdparty/openssl-3.1.1'
make: *** [Makefile:3283: build_sw] Error 2`

**dump：**
`perl configdata.pm --dump

Command line (with current working directory = .):

    /usr/bin/perl ./Configure linux-mips32 --cross-compile-prefix=mips-linux-gnu- --prefix=/home/jamo/workspace/protocol_example/lib CFLAGS=-mabi=32 -mips2 -mmxu3 -mfp64

Perl information:

    /usr/bin/perl
    5.30.0 for x86_64-linux-gnu-thread-multi

Enabled features:

    afalgeng
    aria
    asm
    async
    autoalginit
    autoerrinit
    autoload-config
    bf
    blake2
    bulk
    cached-fetch
    camellia
    capieng
    cast
    chacha
    cmac
    cmp
    cms
    comp
    ct
    deprecated
    des
    dgram
    dh
    dsa
    dso
    dtls
    dynamic-engine
    ec
    ec2m
    ecdh
    ecdsa
    engine
    err
    filenames
    gost
    idea
    legacy
    loadereng
    makedepend
    md4
    mdc2
    module
    multiblock
    nextprotoneg
    ocb
    ocsp
    padlockeng
    pic
    pinshared
    poly1305
    posix-io
    psk
    rc2
    rc4
    rdrand
    rfc3779
    rmd160
    scrypt
    secure-memory
    seed
    shared
    siphash
    siv
    sm2
    sm3
    sm4
    sock
    srp
    srtp
    sse2
    ssl
    ssl-trace
    static-engine
    stdio
    tests
    threads
    tls
    ts
    ui-console
    whirlpool
    tls1
    tls1-method
    tls1_1
    tls1_1-method
    tls1_2
    tls1_2-method
    tls1_3
    dtls1
    dtls1-method
    dtls1_2
    dtls1_2-method

Disabled features:

    acvp-tests          [cascade]        OPENSSL_NO_ACVP_TESTS
    asan                [default]        OPENSSL_NO_ASAN
    buildtest-c++       [default]
    crypto-mdebug       [default]        OPENSSL_NO_CRYPTO_MDEBUG
    devcryptoeng        [default]        OPENSSL_NO_DEVCRYPTOENG
    ec_nistp_64_gcc_128 [default]        OPENSSL_NO_EC_NISTP_64_GCC_128
    egd                 [default]        OPENSSL_NO_EGD
    external-tests      [default]        OPENSSL_NO_EXTERNAL_TESTS
    fips                [default]
    fips-securitychecks [cascade]        OPENSSL_NO_FIPS_SECURITYCHECKS
    fuzz-afl            [default]        OPENSSL_NO_FUZZ_AFL
    fuzz-libfuzzer      [default]        OPENSSL_NO_FUZZ_LIBFUZZER
    ktls                [default]        OPENSSL_NO_KTLS
    md2                 [default]        OPENSSL_NO_MD2 (skip crypto/md2)
    msan                [default]        OPENSSL_NO_MSAN
    rc5                 [default]        OPENSSL_NO_RC5 (skip crypto/rc5)
    sctp                [default]        OPENSSL_NO_SCTP
    trace               [default]        OPENSSL_NO_TRACE
    ubsan               [default]        OPENSSL_NO_UBSAN
    unit-test           [default]        OPENSSL_NO_UNIT_TEST
    uplink              [no uplink_arch] OPENSSL_NO_UPLINK
    weak-ssl-ciphers    [default]        OPENSSL_NO_WEAK_SSL_CIPHERS
    zlib                [default]
    zlib-dynamic        [default]
    ssl3                [default]        OPENSSL_NO_SSL3
    ssl3-method         [default]        OPENSSL_NO_SSL3_METHOD

Config target attributes:

    AR => "ar",
    ARFLAGS => "qc",
    CC => "gcc",
    CFLAGS => "-Wall -O3",
    CXX => "g++",
    CXXFLAGS => "-Wall -O3",
    HASHBANGPERL => "/usr/bin/env perl",
    RANLIB => "ranlib",
    RC => "windres",
    asm_arch => "mips32",
    bn_ops => "BN_LLONG RC4_CHAR",
    build_file => "Makefile",
    build_scheme => [ "unified", "unix" ],
    cflags => "-pthread -mabi=32",
    cppflags => "",
    cxxflags => "-std=c++11 -pthread -mabi=32",
    defines => [ "OPENSSL_BUILDING_OPENSSL" ],
    disable => [  ],
    dso_ldflags => "-Wl,-z,defs",
    dso_scheme => "dlfcn",
    enable => [ "afalgeng" ],
    ex_libs => "-ldl -pthread -latomic",
    includes => [  ],
    lflags => "",
    lib_cflags => "",
    lib_cppflags => "-DOPENSSL_USE_NODELETE",
    lib_defines => [  ],
    module_cflags => "-fPIC",
    module_cxxflags => undef,
    module_ldflags => "-Wl,-znodelete -shared -Wl,-Bsymbolic",
    perl_platform => "Unix",
    perlasm_scheme => "o32",
    shared_cflag => "-fPIC",
    shared_defflag => "-Wl,--version-script=",
    shared_defines => [  ],
    shared_ldflag => "-Wl,-znodelete -shared -Wl,-Bsymbolic",
    shared_rcflag => "",
    shared_sonameflag => "-Wl,-soname=",
    shared_target => "linux-shared",
    thread_defines => [  ],
    thread_scheme => "pthreads",
    unistd => "<unistd.h>",

Recorded environment:

    AR =
    ARFLAGS =
    AS =
    ASFLAGS =
    BUILDFILE =
    CC =
    CFLAGS =
    CPP =
    CPPDEFINES =
    CPPFLAGS =
    CPPINCLUDES =
    CROSS_COMPILE =
    CXX =
    CXXFLAGS =
    HASHBANGPERL =
    LD =
    LDFLAGS =
    LDLIBS =
    MT =
    MTFLAGS =
    OPENSSL_LOCAL_CONFIG_DIR =
    PERL =
    RANLIB =
    RC =
    RCFLAGS =
    RM =
    WINDRES =
    __CNF_CFLAGS =
    __CNF_CPPDEFINES =
    __CNF_CPPFLAGS =
    __CNF_CPPINCLUDES =
    __CNF_CXXFLAGS =
    __CNF_LDFLAGS =
    __CNF_LDLIBS =

Makevars:

    AR              = mips-linux-gnu-ar
    ARFLAGS         = qc
    CC              = mips-linux-gnu-gcc
    CFLAGS          = -mabi=32 -mips2 -mmxu3 -mfp64
    CPPDEFINES      =
    CPPFLAGS        =
    CPPINCLUDES     =
    CROSS_COMPILE   = mips-linux-gnu-
    CXX             = mips-linux-gnu-g++
    CXXFLAGS        = -Wall -O3
    HASHBANGPERL    = /usr/bin/env perl
    LDFLAGS         =
    LDLIBS          =
    PERL            = /usr/bin/perl
    RANLIB          = mips-linux-gnu-ranlib
    RC              = mips-linux-gnu-windres
    RCFLAGS         =

NOTE: These variables only represent the configuration view.  The build file
template may have processed these variables further, please have a look at the
build file for more exact data:
    Makefile

build file:

    Makefile

build file templates:

    Configurations/common0.tmpl
    Configurations/unix-Makefile.tmpl`

**Comment by paulidale** [comment, marked as answer]
Your choice of CFLAGS looks like the problem: `CFLAGS="-mabi=32 -mips2 -mmxu3 -mfp64"`
Some of these appear to be incompatible.
